### ISSUE TITLE
Getting cacheType none while retrievingImage after successful storing. 

### ISSUE THREAD

**Issue by RAITERN** [onevcat/Kingfisher#1174]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [ ] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [ ] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [ ] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

### Issue Description

I am getting cacheType as 'none' while retrieving image for a cacheKey after successful storing of that image. 
Note: For sometimes I can able to retrieveImage, but after a while it returns none. 

I have created my own ImageCache & set the disk expiration time to 'never'. 

```swift
myCache = ImageCache(name: "myCache")
myCache?.diskStorage.config.expiration = .never
```

And store images using corresponding cacheKey for images

```swift
myCache?.store(image, forKey: cacheKey, completionHandler: { (result) in 
    //getting success here 
})
```

I can able to retrieve image for a few minutes. But after sometimes, the same retrieve image func returns cacheType as 'none'. Checked my cacheKey, there is no problem.

```swift
myCache?.retrieveImage(forKey: cacheKey) { result in
    switch result {
        case .success(let value):
            // cacheType is .none                
        case .failure( _):
    }
}
```

Please help.

**Comment by onevcat** [comment]
I tried it but cannot reproduce this issue.

The only possible reasons (if it is not a bug) are: 

- Your disk is full and the system cleared the disk cache.
- Or maybe you are doing some cache cleaning manually.
- Or you stored too many images in the disk and it exceeds the `sizeLimit` and the cleaning task is triggered inside Kingfisher.

Can this issue be reproduced in a sample project?

**Comment by RAITERN** [comment]
I didn't cleared the cache manually. BTW, I have stored many images (nearly 300 icons) but the size of each one will be around 200kb. I didn't set any sizeLimit for disk cache. What is the default sizeLimit for disk cache? Or do I have to explicitly mention the disk sizeLimit.

**Comment by onevcat** [comment]
:( The default value is 0, means no limitation for disk cache.

**Comment by onevcat** [comment]
Does it happen only when you switch your app background and then foreground? Or does it also happens when keeping using your app in foreground?

And since I cannot reproduce it myself, it is not an easy thing to guess what happens. It would be great help if you can provide a stable reproducible sample for this issue.

**Comment by RAITERN** [comment]
Not sure it occurs due to background & foreground switching of the app, because I cannot reproduce all the time. But after a while when I opened the app from background, the images are gone. If there is any solution for this switching case, I could try it.

Sorry I cannot provide you the sample project as it contains my company data. I will try with something similar to provide it to you.

**Comment by RAITERN** [comment]
Yes I can reproduce it now by switching the app to background and then foreground.

**Comment by onevcat** [comment]
Switching app to the background will trigger Kingfisher's auto purge for memory cache. So it means your image is cached to memory successfully, but it fails to be cached to disk for some reasons. Can you try to make some change to your code like below and let me know the result?

```swift
cache.store(image, forKey: key) { result in
    print(result.diskCacheResult)
    print(cache.imageCachedType(forKey: key))
}
```

**Comment by RAITERN** [comment]
As you said, result. diskCacheResult returns ``Result.success(())`` 

And cache.imageCachedType(forKey: key) returns as '.none' if I cleared the memoryCache as in your code. 
If didn't do cache.clearMemoryCache(), then the imageCachedType returns as '.memory'

**Comment by onevcat** [comment]
It is really strange. Let me check.

**Comment by RAITERN** [comment]
Does it have anything to do with my own cache name. If I create a cache with the same name again, it wont create another cache right? 
What my app does is, those images will be downloaded only once while logging in by creating my own imageCache. If I killed the app & open it again, I create the cache by same name and retrieving the images. Only while logging out I'm clearing all the caches.

**Comment by onevcat** [comment]
No problem. The same name will point to the same cache.

Will it work normally if you change to use the default cache `ImageCache.default`?

**Comment by onevcat** [comment]
@RAITERN I've added and temporary log in the branch https://github.com/onevcat/Kingfisher/tree/try/creating-failure 

Can you update your integration to that branch and check the log there? Thanks!

**Comment by RAITERN** [comment]
Just now tried that default.cache and it works good. But I need both my own cache(for disk expiration never) & I'm using default cache for another purpose.

**Comment by onevcat** [comment]
What's the name of your own cache? Is it possible that you are using an invalid name so the disk cache fails?

**Comment by RAITERN** [comment]
"LogoCache" is the name.

`logoCache = ImageCache(name: "LogoCache")`

**Comment by onevcat** [comment]
Seems fine. Maybe try to use all lower case? I am not sure although...

**Comment by onevcat** [comment]
(And I want to declare again that the disk cache is not a good persistence if you want these images live there forever. Kingfisher uses system cache folder for caching purpose, and that folder could be purged by the system without noticing your app or the framework. But it is another story.)

**Comment by RAITERN** [comment]
Hi. Found the issue. On logging out I clear all the folder under cache directory which I use for several purpose to store files, which in turn removes the cache folder created by Kingfisher. However if I login again, it won't create the cache folder again even though I call 'ImageCache(name: "LogoCache")'. That why it cant store images again in disk. 
It would be good to create the cache folder again if not available.

Also is there any way to use these cache image data in app extensions.

**Comment by onevcat** [comment]
It should create the folder again when you call the `ImageCache` initializer, which in turn would call [these code](https://github.com/onevcat/Kingfisher/blob/master/Sources/Cache/DiskStorage.swift#L78-L92). A bit strange why that could happen.
